test/integration/integration.cc

```diff
@@ -813,6 +813,7 @@ void BaseIntegrationTest::testUpstreamProtocolError() {
   IntegrationCodecClientPtr codec_client;
   IntegrationStreamDecoderPtr response(new IntegrationStreamDecoder(*dispatcher_));
   FakeRawConnectionPtr fake_upstream_connection;
+  fake_upstreams_[0]->setWaitForHttpConnection(false);
```

Comment:
We need `FakeUpstream` to be capable of asynchronously waiting for and queuing up arrived connections in `new_connections_` (or the variants I added in this PR).  Previously, the `Network::Connection&` was queued up here, but it was not owned and could be freed on a disconnect. The consumer of `new_connection_` then gets some freed memory. There's no event listener to catch this scenario, that's why I want to wrap the connection with one of these `Fake*Connection` objects that is capable of responding to the disconnect if it happens while queued in `new_connections_`.

Maybe we could get `FakeHttpConnection` to wrap around a `FakeRawConnection`, I can look into this. It seemed to me the common case that folks want to `waitForHttpConnection()` though, only a few care about `waitForRawConnection()`.